benchmarks/query/recording.md

```diff
@@ -0,0 +1,88 @@
+## Summary
+
+Key findings:
+
+- loki without indexes is overall slightly faster than master, except when there are many types
+- loki with indexes is about 2x faster on sites with 10k pages, and 5x faster with 20k pages. But is ever so slightly slower when those pages are split across 100 types.
+
+Overall, loki is a big win for sites with lots of pages of the same type. For smaller sites, the difference is negligible.
```

Comment:
A caching PR I recently worked on had similar findings, which I think is totally reasonable. If it's mostly the same for smaller sites but it makes Gatsby much more scalable I think that's a big win.

I honestly wonder if using lokijs _consistently_ (e.g. for cache and internal Gatsby utilities) would be worthwhile. I'd lean towards yes!

Reply:
The fs fallback is interesting, since I used a similar approach for the cache! Basically I kept 100-200 items in memory, and then it fell back to the filesystem as its cache, which seems to work pretty well for large sites.

Reply:
@Moocar did you measure how much memory is used in these different benchmarks?

@DSchau same question?

Reply:
@KyleAMathews not explicitly (although I can), but I used the upwards bounds of when it ran _out_ of memory as the basis for which I considered that. It is dependent on machine setup, size of items being cached, etc. too.

Reply:
@DSchau Thanks for all the feedback! For caching, using loki would be a step up from using pure arrays, but an even bigger win would probably be to use a real caching library that can also provide persistence. Once that has LRU, paging to disk and other cache-specific semantics, assuming one exists! (I haven't looked)

Reply:
@KyleAMathews Memory usage should be less since we're actually able to remove some caches. But I'll double check. What's your favorite memory profiling tool?

Reply:
That's what I assume too but it's nice to know for sure.

There's probably something on npm but you can also add a setInterval in the code and log out the memory usage every .1 seconds or whatever. You could also log out different events too e.g. when query running starts etc. so you could visualize things easily.